src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/MessagePackHubProtocol.java

```diff
@@ -0,0 +1,580 @@
+            packer.packString(error);
+        } else {
+            packer.packNil();
+        }
+        
+        packer.packBoolean(message.getAllowReconnect());
+        
+        packer.flush();
+        String content = out.toString();
+        out.close();
+        return content;
+    }
+    
+    private Map<String, String> readHeaders(MessageUnpacker unpacker) throws IOException {
+        int headerCount = unpacker.unpackMapHeader();
+        if (headerCount > 0) {
+            Map<String, String> headers = new HashMap<String, String>();
+            for (int i = 0; i < headerCount; i++) {
+                headers.put(unpacker.unpackString(), unpacker.unpackString());
+            }
+            return headers;
+        } else {
+            return null;
+        }
+    }
+
+    private void writeHeaders(Map<String, String> headers, MessagePacker packer) throws IOException {
+        if (headers != null) {
+            packer.packMapHeader(headers.size());
+            for (String k: headers.keySet()) {
+                packer.packString(k);
+                packer.packString(headers.get(k));
+            }    
+        } else {
+            packer.packMapHeader(0);
+        }
+    }
+    
+    private Collection<String> readStreamIds(MessageUnpacker unpacker) throws IOException {
+        int streamCount = unpacker.unpackArrayHeader();
+        Collection<String> streams = null;
+        
+        if (streamCount > 0) {
+            streams = new ArrayList<String>();
+            for (int i = 0; i < streamCount; i++) {
+                streams.add(unpacker.unpackString());
+            }
+        }
+        
+        return streams;
+    }
+    
+    private void writeStreamIds(Collection<String> streamIds, MessagePacker packer) throws IOException {
+        if (streamIds != null) {
+            packer.packArrayHeader(streamIds.size());
+            for (String s: streamIds) {
+                packer.packString(s);
+            }
+        } else {
+            packer.packArrayHeader(0);
+        }
+    }
+    
+    private Object[] bindArguments(MessageUnpacker unpacker, List<Class<?>> paramTypes) throws IOException {
+        int argumentCount = unpacker.unpackArrayHeader();
+        
+        if (paramTypes.size() != argumentCount) {
+            throw new RuntimeException(String.format("Invocation provides %d argument(s) but target expects %d.", argumentCount, paramTypes.size()));
+        }
+        
+        Object[] arguments = new Object[argumentCount];
+        
+        for (int i = 0; i < argumentCount; i++) {
+            arguments[i] = readValue(unpacker, paramTypes.get(i));
+        }
+        
+        return arguments;
+    }
+    
+    private Object readValue(MessageUnpacker unpacker, Class<?> itemType) throws IOException {
+        // This shouldn't ever get called with itemType == null, but we return anyways to avoid NullPointerExceptions
+        if (itemType == null) {
```

Comment:
Should I even account for itemType being null? In theory it should never happen (would indicate that a function took an argument of type null, or returned something of type null, which shouldn't be possible) but the test InvocationBinder returns null: https://github.com/dotnet/aspnetcore/blob/0d156487aa27df65d68fc7fa31e7b6cb9f4d7581/src/SignalR/clients/java/signalr/src/test/java/com/microsoft/signalr/JsonHubProtocolTest.java#L286

Reply:
Could this happen given a protocol violation by the server?

If so, we should at the very least log an error and fail the corresponding client-to-server invocation or skip the corresponding server-to-client invocation. Closing the entire HubConnection with an error might be even better.

If this actually cannot happen short of memory corruption or similar, just assert.

Reply:
It could happen if the server provides an InvocationBinder that returns `null` from `getReturnType(invocationId)` when sending a StreamItem (or if the client does the same when sending an invocation)